third_party/move/move-vm/runtime/src/loader.rs

```diff
@@ -745,11 +742,128 @@ impl Loader {
             .collect::<PartialVMResult<Vec<_>>>()
             .map_err(|err| err.finish(Location::Undefined))?;
 
-        // verify type arguments
+        Ok((module, func, parameters, return_))
+    }
+
+    // Matches the actual returned type to the expected type, binding any type args to the
+    // necessary type as stored in the map. Returns true if a successful match is made.
+    fn match_return_type<'a>(
+        returned: &Type,
+        expected: &'a Type,
+        map: &mut BTreeMap<usize, &'a Type>,
+    ) -> bool {
+        match (returned, expected) {
+            // The important case, deduce the type params
+            (Type::TyParam(idx), _) => {
+                if let Option::Some(bounded_type) = map.get(idx) {
+                    // The type argument was already bound, so to be consistent they have to equal
+                    *bounded_type == expected
+                } else {
+                    // In order for the returned type to match the expected type this type parameter
+                    // must be expected
+                    map.insert(*idx, expected);
+                    true
+                }
+            },
+            // Recursive types we need to recurse the matching types
+            (Type::Reference(ret_inner), Type::Reference(expected_inner))
+            | (Type::MutableReference(ret_inner), Type::MutableReference(expected_inner))
+            | (Type::Vector(ret_inner), Type::Vector(expected_inner)) => {
+                Self::match_return_type(ret_inner, expected_inner, map)
+            },
+            // For struct instantiations we need to match all fields
+            (
+                Type::StructInstantiation(ret_idx, ret_fields),
+                Type::StructInstantiation(expected_idx, expected_fields),
+            ) => {
+                *ret_idx == *expected_idx
+                    && ret_fields.len() == expected_fields.len()
+                    && ret_fields
+                        .iter()
+                        .zip(expected_fields.iter())
+                        .all(|types| Self::match_return_type(types.0, types.1, map))
+            },
+            // For the rest we need to assure the types match
+            _ => std::mem::discriminant(returned) == std::mem::discriminant(expected),
```

Comment:
I strongly recommend you write out the remaining pairs instead
```
    (U8, U8) | (U16, U16) | (U32, U32) => true,
    _ => false,
```
Functions in std::mem are low level memory operations, and Rust generally encourages programmers to stay abstract if possible. Using `std::mem::discriminant` implies specific memory layouts, which IMO isn't something this function should care about.

Reply:
discriminant uses impl details under the hood yes, but it is specifically meant for cases like this. It returns a opaque type that can be compared only as a well defined discriminant

Reply:
Personally I like brevity, but the case of compiler catches is a strong one. Unfortunately without using _ we get a NxN mix and with it the compiler wouldn't catch it either.

Reply:
Yeah, that's the tradeoff we have to weight. Usually we write out match cases that are O(N), but not the ones that are O(N^2), because that's simply too much.

Reply:
TyParam is explicitly matched in the first pattern, it's actually the most important case.

Reply:
Could we also get this fixed before landing? The discriminant comparison does seem vulnerable to me, in case new variants are introduced upstream.